File: zodiacview/src/main/java/com/dev/tanakov/zodiacview/ZodiacView.java
/**
 * Copyright 10/20/16 Vladimir Tanakov
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.dev.tanakov.zodiacview;

import java.util.ArrayList;
import java.util.List;

import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.util.AttributeSet;
import android.view.View;

public class ZodiacView extends View {

    //region DEFAULT VALUES

    private static final int DEFAULT_STAR_COUNT = 30;
    private static final int DEFAULT_STAR_SIZE_MIN = 10;
    private static final int DEFAULT_STAR_SIZE_MAX = 20;
    private static final int DEFAULT_RELATION_SIZE = 5;
    private static final float DEFAULT_SPEED = 0.7f;
    private static final int DEFAULT_DISTANCE = 200;
    private static final String DEFAULT_COLOR_BACKGROUND = "#16151f";
    private static final String DEFAULT_COLOR_STAR = "#49348b";
    private static final String DEFAULT_COLOR_RELATION = "#49348b";

    //endregion

    private int mWidth;
    private int mHeight;

    private List<Star> mStars = new ArrayList<>();
    private int mStarCount = DEFAULT_STAR_COUNT;
    private int mStarSizeMin = DEFAULT_STAR_SIZE_MIN;
    private int mStarSizeMax = DEFAULT_STAR_SIZE_MAX;
    private int mRelationSize = DEFAULT_RELATION_SIZE;

    private float mSpeed = DEFAULT_SPEED;
    private int mDistance = DEFAULT_DISTANCE;

    private Paint mPaintRelation;
    private Paint mPaintStar;

    private int mColorBackground = Color.parseColor(DEFAULT_COLOR_BACKGROUND);
    private int mColorStar = Color.parseColor(DEFAULT_COLOR_STAR);
    private int mColorRelation = Color.parseColor(DEFAULT_COLOR_RELATION);

    public ZodiacView(Context context) {
        super(context);
        init();
    }

    public ZodiacView(Context context, AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public ZodiacView(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);

        TypedArray a = context.obtainStyledAttributes(attrs, R.styleable.ZodiacView, defStyleAttr, 0);

        mStarCount = a.getInt(R.styleable.ZodiacView_zv_star_count, DEFAULT_STAR_COUNT);
        mStarSizeMin = a.getDimensionPixelSize(R.styleable.ZodiacView_zv_star_size_min, DEFAULT_STAR_SIZE_MIN);
        mStarSizeMax = a.getDimensionPixelSize(R.styleable.ZodiacView_zv_star_size_max, DEFAULT_STAR_SIZE_MAX);
        mRelationSize = a.getDimensionPixelSize(R.styleable.ZodiacView_zv_relation_size, DEFAULT_RELATION_SIZE);
        mSpeed = a.getFloat(R.styleable.ZodiacView_zv_speed, DEFAULT_SPEED);
        mDistance = a.getInt(R.styleable.ZodiacView_zv_distance, DEFAULT_DISTANCE);
        mColorBackground = a.getColor(R.styleable.ZodiacView_zv_color_bg, Color.parseColor(DEFAULT_COLOR_BACKGROUND));
        mColorStar = a.getColor(R.styleable.ZodiacView_zv_color_star, Color.parseColor(DEFAULT_COLOR_STAR));
        mColorRelation = a.getColor(R.styleable.ZodiacView_zv_color_relation, Color.parseColor(DEFAULT_COLOR_RELATION));

        a.recycle();

        init();
    }

    @Override
    protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
        super.onMeasure(widthMeasureSpec, heightMeasureSpec);
        mWidth = View.MeasureSpec.getSize(widthMeasureSpec);
        mHeight = View.MeasureSpec.getSize(heightMeasureSpec);
        createStars();
    }

    @Override
    protected void onDraw(Canvas canvas) {
        super.onDraw(canvas);

        for (int i = 0; i < mStarCount; i++) {
            mStars.get(i).connectedStars.clear();
        }

        for (int i = 0; i < mStarCount; i++) {
            Star star = mStars.get(i);
            moveStar(star);

            canvas.drawCircle(star.x, star.y, star.size, mPaintStar);

            for (int j = 0; j < mStarCount; j++) {
                Star refStar = mStars.get(j);
                if (star == refStar || refStar.connectedStars.contains(star)) {
                    continue;
                }

                if (Math.abs(star.x - refStar.x) < mDistance && Math.abs(star.y - refStar.y) < mDistance) {
                    star.connectedStars.add(refStar);
                    drawRelation(canvas, star, refStar);
                }
            }
        }

        invalidate();
    }

    private void init() {
        setBackgroundColor(mColorBackground);

        mPaintStar = new Paint();
        mPaintStar.setColor(mColorStar);
        mPaintStar.setStyle(Paint.Style.FILL);
        mPaintStar.setAntiAlias(true);

        mPaintRelation = new Paint();
        mPaintRelation.setColor(mColorRelation);
        mPaintRelation.setStrokeWidth(mRelationSize);
        mPaintRelation.setStyle(Paint.Style.FILL);
        mPaintRelation.setAntiAlias(true);
    }

    private void createStars() {
        mStars.clear();
        for (int i = 0; i < mStarCount; i++) {
            mStars.add(createStar());
        }
    }

    private Star createStar() {
        Star star = new Star();
        star.x = (float) Math.random() * mWidth;
        star.y = (float) Math.random() * mHeight;
        star.dirX = (float) (Math.random() - 0.5);
        star.dirY = (float) (Math.random() - 0.5);
        star.size = (float) (Math.random() * (mStarSizeMax - mStarSizeMin)) + mStarSizeMin;

        return star;
    }

    private Star moveStar(Star star) {
        star.x += star.dirX * mSpeed;
        star.y += star.dirY * mSpeed;

        if (star.y + star.size > mHeight) {
            star.dirY = (float) (Math.random() * -1);
        } else if (star.y - star.size < 0) {
            star.dirY = (float) Math.random();
        } else if (star.x + star.size > mWidth) {
            star.dirX = (float) Math.random() * -1;
        } else if (star.x - star.size < 0) {
            star.dirX = (float) Math.random();
        }

        return star;
    }

    private void drawRelation(Canvas canvas, Star star, Star refStar) {
        double x = Math.abs(star.x - refStar.x);
        double y = Math.abs(star.y - refStar.y);
        double dist = Math.sqrt(Math.pow(x, 2) + Math.pow(y, 2));
        int opacity = (int) ((1.4 - dist / mDistance) * 255);

        mPaintRelation.setAlpha(opacity);

        canvas.drawLine(star.x, star.y, refStar.x, refStar.y, mPaintRelation);
    }

    //region SETTERS

    public void setStarCount(int count) {
        mStarCount = count;
    }

    public void setStarSizeMin(int size) {
        mStarSizeMin = size;
    }

    public void setStarSizeMax(int size) {
        mStarSizeMax = size;
    }

    public void setRelationSize(int size) {
        mRelationSize = size;
    }

    public void setSpeed(float speed) {
        mSpeed = speed;
    }

    public void setDistance(int distance) {
        mDistance = distance;
    }

    public void setColorBackground(int bgColor) {
        mColorBackground = bgColor;
    }

    public void setColorStar(int starColor) {
        mColorStar = starColor;
    }

    public void setColorRelation(int relationColor) {
        mColorRelation = relationColor;
    }

    //endregion
}

File: zodiacview/src/main/java/com/dev/tanakov/zodiacview/ZodiacView.kt
/**
 * Copyright 16/08/21 Vladimir Tanakov
 *
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.dev.tanakov.zodiacview

import android.content.Context
import android.graphics.Canvas
import android.graphics.Paint
import android.util.AttributeSet
import android.view.MotionEvent
import android.view.View
import android.view.View.OnTouchListener
import androidx.core.content.ContextCompat
import kotlin.math.abs
import kotlin.math.pow
import kotlin.math.sqrt

class ZodiacView : View, OnTouchListener {

    private var viewWidth = 0
    private var viewHeight = 0
    private val stars: MutableList<Star> = mutableListOf()

    private var starCount = DEFAULT_STAR_COUNT
    private var starSizeMin = DEFAULT_STAR_SIZE_MIN
    private var starSizeMax = DEFAULT_STAR_SIZE_MAX
    private var relationSize = DEFAULT_RELATION_SIZE
    private var speed = DEFAULT_SPEED
    private var distance = DEFAULT_DISTANCE
    private var colorBackground = ContextCompat.getColor(context, R.color.zv_default_color_background)
    private var colorStar = ContextCompat.getColor(context, R.color.zv_default_color_star)
    private var colorRelation = ContextCompat.getColor(context, R.color.zv_default_color_relation)
    private var interactionEnabled = DEFAULT_INTERACTION_ENABLED

    private lateinit var paintStar: Paint
    private lateinit var paintRelation: Paint

    private constructor(context: Context) : super(context)

    @JvmOverloads
    constructor(context: Context, attrs: AttributeSet?, defStyleAttr: Int = 0) : super(context, attrs, defStyleAttr) {
        val a = context.obtainStyledAttributes(attrs, R.styleable.ZodiacView, defStyleAttr, 0)
        starCount = a.getInt(R.styleable.ZodiacView_zv_star_count, DEFAULT_STAR_COUNT)
        starSizeMin = a.getDimensionPixelSize(R.styleable.ZodiacView_zv_star_size_min, DEFAULT_STAR_SIZE_MIN)
        starSizeMax = a.getDimensionPixelSize(R.styleable.ZodiacView_zv_star_size_max, DEFAULT_STAR_SIZE_MAX)
        relationSize = a.getDimensionPixelSize(R.styleable.ZodiacView_zv_relation_size, DEFAULT_RELATION_SIZE)
        speed = a.getFloat(R.styleable.ZodiacView_zv_speed, DEFAULT_SPEED)
        distance = a.getInt(R.styleable.ZodiacView_zv_distance, DEFAULT_DISTANCE)
        colorBackground = a.getColor(
            R.styleable.ZodiacView_zv_color_bg,
            ContextCompat.getColor(context, R.color.zv_default_color_background)
        )
        colorStar = a.getColor(
            R.styleable.ZodiacView_zv_color_star,
            ContextCompat.getColor(context, R.color.zv_default_color_star)
        )
        colorRelation = a.getColor(
            R.styleable.ZodiacView_zv_color_relation,
            ContextCompat.getColor(context, R.color.zv_default_color_relation)
        )
        interactionEnabled = a.getBoolean(R.styleable.ZodiacView_zv_interaction_enabled, DEFAULT_INTERACTION_ENABLED)
        a.recycle()
        init()
    }

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        super.onMeasure(widthMeasureSpec, heightMeasureSpec)
        viewWidth = MeasureSpec.getSize(widthMeasureSpec)
        viewHeight = MeasureSpec.getSize(heightMeasureSpec)
        createStars()
    }

    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)
        val starCount = stars.size
        for (i in 0 until starCount) {
            stars[i].connectedStars.clear()
        }
        for (i in 0 until starCount) {
            val star = stars[i]
            moveStar(star)
            canvas.drawCircle(star.x, star.y, star.size, paintStar)
            for (j in 0 until starCount) {
                val refStar = stars[j]
                if (star === refStar || refStar.connectedStars.contains(star)) {
                    continue
                }
                if (abs(star.x - refStar.x) < distance && abs(star.y - refStar.y) < distance) {
                    star.connectedStars.add(refStar)
                    drawRelation(canvas, star, refStar)
                }
            }
        }
        invalidate()
    }

    override fun onTouch(view: View, motionEvent: MotionEvent): Boolean {
        val x = motionEvent.x
        val y = motionEvent.y
        when (motionEvent.action) {
            MotionEvent.ACTION_DOWN -> {
                val star = createStar().apply {
                    this.x = x
                    this.y = y
                }
                stars.add(star)
            }
        }
        return true
    }

    private fun init() {
        setBackgroundColor(colorBackground)
        paintStar = Paint().apply {
            color = colorStar
            style = Paint.Style.FILL
            isAntiAlias = true
        }
        paintRelation = Paint().apply {
            color = colorRelation
            strokeWidth = relationSize.toFloat()
            style = Paint.Style.FILL
            isAntiAlias = true
        }
        if (interactionEnabled) {
            setOnTouchListener(this)
        }
    }

    private fun createStars() {
        stars.clear()
        for (i in 0 until starCount) {
            stars.add(createStar())
        }
    }

    private fun createStar(): Star {
        val star = Star()
        star.x = Math.random().toFloat() * viewWidth
        star.y = Math.random().toFloat() * viewHeight
        star.dirX = (Math.random() - 0.5).toFloat()
        star.dirY = (Math.random() - 0.5).toFloat()
        star.size = (Math.random() * (starSizeMax - starSizeMin)).toFloat() + starSizeMin
        return star
    }

    private fun moveStar(star: Star): Star {
        star.x += star.dirX * speed
        star.y += star.dirY * speed
        when {
            star.y + star.size > viewHeight -> star.dirY = (Math.random() * -1).toFloat()
            star.y - star.size < 0          -> star.dirY = Math.random().toFloat()
            star.x + star.size > viewWidth  -> star.dirX = Math.random().toFloat() * -1
            star.x - star.size < 0          -> star.dirX = Math.random().toFloat()
        }
        return star
    }

    private fun drawRelation(canvas: Canvas, star: Star, refStar: Star) {
        val x = abs(star.x - refStar.x)
        val y = abs(star.y - refStar.y)
        val dist = sqrt(x.pow(2) + y.pow(2))
        val opacity = (1.4 - dist / distance) * 255
        paintRelation.alpha = opacity.toInt()
        canvas.drawLine(star.x, star.y, refStar.x, refStar.y, paintRelation)
    }

    companion object {
        fun setup(context: Context, config: ZodiacConfig): ZodiacView {
            return ZodiacView(context).apply {
                starCount = config.starCount
                starSizeMin = config.starSizeMin
                starSizeMax = config.starSizeMax
                relationSize = config.relationSize
                speed = config.speed
                distance = config.distance
                interactionEnabled = config.interactionEnabled
                colorBackground = ContextCompat.getColor(context, config.colorBackground)
                colorStar = ContextCompat.getColor(context, config.colorStar)
                colorRelation = ContextCompat.getColor(context, config.colorRelation)
                init()
            }
        }
    }
}
